File: sprint6/Java/lesson4code3.java
public void DFS(int startVertex) {
    Stack<Integer> stack = new Stack<>();
    stack.push(startVertex);
    while (!stack.empty()) {
        int v = stack.pop(); // Получаем из стека очередную вершину.
        if (color[v] == white) {
            // Красим вершину в серый. И сразу кладём её обратно в стек:
            // это позволит алгоритму позднее вспомнить обратный путь по графу.
            color[v] = gray;
            stack.push(v);
            // Теперь добавляем в стек все непосещённые соседние вершины,
            // вместо вызова рекурсии
            for (int w : adjList[v]) { // Перебираем смежные вершины.
                if (color[w] == white) { // Если вершина не посещена, то
                    stack.push(w);
                }
            }
        } else if (color[v] == gray) {
            // Серую вершину мы могли получить из стека только на обратном пути.
            // Следовательно, её следует перекрасить в чёрный.
            color[v] = black;
        }
    }
}

File: sprint6/By lessons/lesson4code3/lesson4code3.swift
func DFS(_ startVertex: Int, adjList: [[Int]], color: inout [Color]) {
    var st = [Int]()
    st.append(startVertex)
    
    while !st.isEmpty {
        let v = st.last!  // Получаем из стека очередную вершину.
        st.removeLast()
        
        if color[v] == .white {
            // Красим вершину в серый. И сразу кладём её обратно в стек:
            // это позволит алгоритму позднее вспомнить обратный путь по графу.
            color[v] = .gray
            st.append(v)
            
            // Теперь добавляем в стек все непосещённые соседние вершины,
            // вместо вызова рекурсии
            for w in adjList[v] {  // Перебираем смежные вершины.
                if color[w] == .white {  // Если вершина не посещена, то
                    st.append(w)
                }
            }
        } else if color[v] == .gray {
            // Серую вершину мы могли получить из стека только на обратном пути.
            // Следовательно, её следует перекрасить в чёрный.
            color[v] = .black
        }
    }
}
